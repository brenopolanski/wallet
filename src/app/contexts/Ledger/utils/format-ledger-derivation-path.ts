import { LedgerDerivationScheme } from "app/contexts/Ledger/contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
