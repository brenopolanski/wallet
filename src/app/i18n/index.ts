import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// Helpers
import { buildTranslations } from "./helpers";

const defaultNS = "translation";

const resources = {
	en: {
		translation: buildTranslations(),
	},
};

// eslint-disable-next-line
i18n.use(initReactI18next).init({
	defaultNS,
	lng: "en",
	ns: [defaultNS],
	resources,
});

export { defaultNS, i18n, resources };
