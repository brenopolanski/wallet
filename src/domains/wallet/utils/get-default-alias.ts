import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let sameCoinWalletsCount = sameCoinWallets.length || 1;

	while (profile.wallets().findByAlias(makeAlias(sameCoinWalletsCount))) {
		sameCoinWalletsCount++;
	}

	return makeAlias(sameCoinWalletsCount);
};
