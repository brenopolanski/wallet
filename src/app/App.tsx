// import { TRX } from "@payvo/sdk-trx";
// import { XLM } from "@payvo/sdk-xlm";
// import { XRP } from "@payvo/sdk-xrp";
// import { ZIL } from "@payvo/sdk-zil";
import "focus-visible";

import LedgerTransportNodeHID from "@ledgerhq/hw-transport-node-hid-singleton";
// import { LUNA } from "@payvo/sdk-luna";
// import { NANO } from "@payvo/sdk-nano";
// import { NEO } from "@payvo/sdk-neo";
import { Environment } from "@payvo/profiles";
// import { ADA } from "@payvo/sdk-ada";
import { ARK } from "@payvo/sdk-ark";
// import { ATOM } from "@payvo/sdk-atom";
// import { AVAX } from "@payvo/sdk-avax";
// import { BTC } from "@payvo/sdk-btc";
// import { DOT } from "@payvo/sdk-dot";
// import { EGLD } from "@payvo/sdk-egld";
// import { ETH } from "@payvo/sdk-eth";
import { LSK } from "@payvo/sdk-lsk";
import { Offline } from "domains/error/pages";
import { ExchangeProvider } from "domains/exchange/contexts/Exchange";
import { Splash } from "domains/splash/pages";
import React, { useLayoutEffect, useState } from "react";
import { useErrorHandler } from "react-error-boundary";
import { I18nextProvider, useTranslation } from "react-i18next";
import { useHistory } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import { middlewares, RouterView, routes } from "router";
import { StubStorage } from "tests/mocks";
import { setThemeSource, shouldUseDarkColors } from "utils/electron-utils";
import { bootEnvWithProfileFixtures, isE2E, isUnit } from "utils/test-helpers";

import { SyncErrorMessage } from "./components/ProfileSyncStatusMessage";
import { ConfigurationProvider, EnvironmentProvider, LedgerProvider, useEnvironmentContext } from "./contexts";
import { useDeeplink, useNetworkStatus, useProfileSynchronizer } from "./hooks";
import { i18n as index18n } from "./i18n";
import { PluginProviders } from "./PluginProviders";
import { SentryProvider } from "./sentry/SentryProvider";
import { SentryRouterWrapper } from "./sentry/SentryRouterWrapper";
import { httpClient, toasts } from "./services";

const RouteWrappers = ({ children }: { children: React.ReactNode }) => (
	<SentryRouterWrapper>{children}</SentryRouterWrapper>
);

const Main = () => {
	const [showSplash, setShowSplash] = useState(true);
	const { env } = useEnvironmentContext();
	const isOnline = useNetworkStatus();
	const history = useHistory();
	const { t } = useTranslation();

	useProfileSynchronizer({
		onProfileRestoreError: () => history.push("/"),
		onProfileSignOut: () => {
			toasts.dismiss();
		},
		onProfileSyncComplete: async () => {
			await toasts.dismiss();
			toasts.success(t("COMMON.PROFILE_SYNC_COMPLETED"));
		},
		onProfileSyncError: async (failedNetworkNames, retryProfileSync) => {
			await toasts.dismiss();

			toasts.warning(
				<SyncErrorMessage
					failedNetworkNames={failedNetworkNames}
					onRetry={async () => {
						await toasts.dismiss();
						retryProfileSync();
					}}
				/>,
			);
		},
		onProfileSyncStart: () => {
			toasts.warning(t("COMMON.PROFILE_SYNC_STARTED"), { autoClose: false });
		},
	});

	useDeeplink();

	useLayoutEffect(() => {
		setThemeSource("system");

		document.body.classList.remove(`theme-${shouldUseDarkColors() ? "light" : "dark"}`);
		document.body.classList.add(`theme-${shouldUseDarkColors() ? "dark" : "light"}`);
	}, []);

	const handleError = useErrorHandler();

	useLayoutEffect(() => {
		const boot = async () => {
			try {
				/* istanbul ignore next */
				if (isE2E() || isUnit()) {
					await bootEnvWithProfileFixtures({ env, shouldRestoreDefaultProfile: isUnit() });

					setShowSplash(false);
					return;
				}

				/* istanbul ignore next */
				await env.verify();
				/* istanbul ignore next */
				await env.boot();
			} catch (error) {
				handleError(error);
			}

			setShowSplash(false);
		};

		boot();
	}, [env, handleError]);

	const renderContent = () => {
		if (showSplash) {
			return <Splash />;
		}

		if (!isOnline) {
			return <Offline />;
		}

		return <RouterView routes={routes} middlewares={middlewares} wrapper={RouteWrappers} />;
	};

	return (
		<main data-testid="Main">
			<ToastContainer closeOnClick={false} newestOnTop />

			{renderContent()}
		</main>
	);
};

export const App = () => {
	/**
	 * Ensure that the Environment object will not be recreated when the state changes,
	 * as the data is stored in memory by the `DataRepository`.
	 */

	/* istanbul ignore next */
	const storage = isE2E() || isUnit() ? new StubStorage() : "indexeddb";

	const [environment] = useState(
		() =>
			new Environment({
				coins: {
					// ADA,
					ARK,
					// ATOM,
					// AVAX,
					// BTC,
					// DOT,
					// ETH,
					// EGLD,
					LSK,
					// NEO,
					// NANO,
					// LUNA,
					// TRX,
					// XLM,
					// XRP,
					// ZIL,
				},
				httpClient,
				storage,
			}),
	);

	return (
		<I18nextProvider i18n={index18n}>
			<EnvironmentProvider env={environment}>
				<ConfigurationProvider defaultConfiguration={{ profileIsSyncingExchangeRates: true }}>
					<SentryProvider>
						<LedgerProvider transport={LedgerTransportNodeHID}>
							<PluginProviders>
								<ExchangeProvider>
									<Main />
								</ExchangeProvider>
							</PluginProviders>
						</LedgerProvider>
					</SentryProvider>
				</ConfigurationProvider>
			</EnvironmentProvider>
		</I18nextProvider>
	);
};
