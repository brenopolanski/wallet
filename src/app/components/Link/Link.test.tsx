import { buildTranslations } from "app/i18n/helpers";
import { toasts } from "app/services";
import electron from "electron";
import React from "react";
import { fireEvent, render, screen } from "utils/testing-library";

import { Link } from "./Link";

const translations = buildTranslations();

describe("Link", () => {
	it("should render", () => {
		const { asFragment } = render(<Link to="/test">Test</Link>);

		expect(screen.getByTestId("Link")).toHaveTextContent("Test");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render external", () => {
		render(
			<Link to="https://payvo.com/" isExternal>
				ARK.io
			</Link>,
		);

		expect(screen.getByTestId("Link")).toHaveAttribute("rel", "noopener noreferrer");
		expect(screen.getByTestId("Link__external")).toBeInTheDocument();
	});

	it("should render external without children", () => {
		const { asFragment } = render(<Link to="https://payvo.com" isExternal />);

		expect(screen.getByTestId("Link__external")).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should open an external link", () => {
		const ipcRendererMock = jest.spyOn(electron.ipcRenderer, "send").mockImplementation();

		const externalLink = "https://payvo.com/";

		const { asFragment } = render(<Link to={externalLink} isExternal />);

		fireEvent.click(screen.getByTestId("Link"));

		expect(ipcRendererMock).toHaveBeenCalledWith("open-external", externalLink);
		expect(asFragment()).toMatchSnapshot();
	});

	it("should show a toast when trying to open an invalid external link", () => {
		const externalLink = "invalid-url";

		const toastSpy = jest.spyOn(toasts, "error");

		const { asFragment } = render(<Link to={externalLink} isExternal />);

		fireEvent.click(screen.getByTestId("Link"));

		expect(toastSpy).toHaveBeenCalledWith(translations.COMMON.ERRORS.INVALID_URL.replace("{{url}}", "invalid-url"));
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with tooltip", () => {
		const { asFragment, baseElement } = render(
			<Link to="/test" tooltip="Custom Tooltip">
				Test
			</Link>,
		);
		const link = screen.getByTestId("Link");

		fireEvent.mouseEnter(link);

		expect(baseElement).toHaveTextContent("Custom Tooltip");

		fireEvent.click(link);

		expect(asFragment()).toMatchSnapshot();
	});
});
