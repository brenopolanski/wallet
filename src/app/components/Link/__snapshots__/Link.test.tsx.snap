// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should open an external link 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-ec014t-0 eeKUfa group ring-focus"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="#"
    rel="noopener noreferrer"
    to="https://payvo.com/"
  >
    <span
      class="Link__Content-sc-ec014t-1 fPOlTq"
    />
    <div
      class="sc-bdvvtL fCrhkg flex-shrink-0 duration-200"
      data-testid="Link__external"
      height="16"
      width="16"
    >
      <svg>
        arrow-external.svg
      </svg>
    </div>
  </a>
</DocumentFragment>
`;

exports[`Link should render 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-ec014t-0 eeKUfa group ring-focus"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="/test"
  >
    <span
      class="Link__Content-sc-ec014t-1 fPOlTq"
    >
      Test
    </span>
  </a>
</DocumentFragment>
`;

exports[`Link should render external without children 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-ec014t-0 eeKUfa group ring-focus"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="#"
    rel="noopener noreferrer"
    to="https://payvo.com"
  >
    <span
      class="Link__Content-sc-ec014t-1 fPOlTq"
    />
    <div
      class="sc-bdvvtL fCrhkg flex-shrink-0 duration-200"
      data-testid="Link__external"
      height="16"
      width="16"
    >
      <svg>
        arrow-external.svg
      </svg>
    </div>
  </a>
</DocumentFragment>
`;

exports[`Link should render with tooltip 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-ec014t-0 eeKUfa group ring-focus"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="/test"
  >
    <span
      class="Link__Content-sc-ec014t-1 fPOlTq"
    >
      Test
    </span>
  </a>
</DocumentFragment>
`;

exports[`Link should show a toast when trying to open an invalid external link 1`] = `
<DocumentFragment>
  <a
    class="Link__AnchorStyled-sc-ec014t-0 eeKUfa group ring-focus"
    data-ring-focus-margin="-m-1"
    data-testid="Link"
    href="#"
    rel="noopener noreferrer"
    to="invalid-url"
  >
    <span
      class="Link__Content-sc-ec014t-1 fPOlTq"
    />
    <div
      class="sc-bdvvtL fCrhkg flex-shrink-0 duration-200"
      data-testid="Link__external"
      height="16"
      width="16"
    >
      <svg>
        arrow-external.svg
      </svg>
    </div>
  </a>
</DocumentFragment>
`;
